File: src/main/java/info/szadkowski/contact/configuration/ThrottlerConfiguration.java
package info.szadkowski.contact.configuration;

import info.szadkowski.contact.properties.ThrottlingProperties;
import info.szadkowski.contact.throttle.Throttler;
import info.szadkowski.contact.throttle.ThrottlerFactory;
import info.szadkowski.contact.throttle.properties.ThrottleConfiguration;
import info.szadkowski.contact.throttle.time.SystemTimeProvider;
import info.szadkowski.contact.throttle.time.TimeProvider;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.scheduling.TaskScheduler;

import java.util.List;

@Configuration
public class ThrottlerConfiguration {

  @Autowired
  private ThrottlingProperties properties;

  @Bean
  public TimeProvider timeProvider() {
    return new SystemTimeProvider();
  }

  @Bean
  public ThrottlerFactory throttlerFactory(TimeProvider timeProvider) {
    return new ThrottlerFactory(timeProvider);
  }

  @Bean
  public Throttler ipThrottler(ThrottlerFactory throttlerFactory) {
    return createThrottler(throttlerFactory, properties.getIp());
  }

  @Bean
  public Throttler allThrottler(ThrottlerFactory throttlerFactory) {
    return createThrottler(throttlerFactory, properties.getAll());
  }

  @Bean(initMethod = "run")
  public Runnable clearer(TaskScheduler taskScheduler,
                          List<Throttler> throttlers) {
    return () -> {
      for (Throttler throttler : throttlers) {
        taskScheduler.scheduleAtFixedRate(throttler::clearExpired, properties.getClearExpiredRate());
      }
    };
  }

  private Throttler createThrottler(ThrottlerFactory throttlerFactory, ThrottlingProperties.ThrottlingScope scope) {
    return throttlerFactory.create(
            scope.getWindow(),
            ThrottleConfiguration.builder().limit(scope.getLimit()).build()
    );
  }
}

File: src/main/kotlin/info/szadkowski/contact/configuration/ThrottlerConfiguration.kt
package info.szadkowski.contact.configuration

import info.szadkowski.contact.properties.ThrottlingProperties
import info.szadkowski.contact.throttle.Throttler
import info.szadkowski.contact.throttle.ThrottlerFactory
import info.szadkowski.contact.throttle.properties.ThrottleConfiguration
import info.szadkowski.contact.throttle.time.SystemTimeProvider
import info.szadkowski.contact.throttle.time.TimeProvider
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.scheduling.TaskScheduler

@Configuration
class ThrottlerConfiguration {

    @Autowired
    lateinit var properties: ThrottlingProperties

    @Bean
    fun timeProvider() = SystemTimeProvider()

    @Bean
    fun throttlerFactory(timeProvider: TimeProvider) = ThrottlerFactory(timeProvider)

    @Bean
    fun ipThrottler(throttlerFactory: ThrottlerFactory) = createThrottler(throttlerFactory, properties.ip!!)

    @Bean
    fun allThrottler(throttlerFactory: ThrottlerFactory) = createThrottler(throttlerFactory, properties.all!!)

    @Bean(initMethod = "run")
    fun clearer(taskScheduler: TaskScheduler, throttlers: List<Throttler>) = Runnable {
        for (throttler in throttlers) {
            taskScheduler.scheduleAtFixedRate(throttler::clearExpired, properties.clearExpiredRate)
        }
    }

    private fun createThrottler(throttlerFactory: ThrottlerFactory, scope: ThrottlingProperties.ThrottlingScope) =
        throttlerFactory.create(
            scope.window,
            ThrottleConfiguration(limit = scope.limit)
        )
}
